pkg/webhook/resourcedistribution/validating/unifiedresource.go

```diff
@@ -0,0 +1,156 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/kubernetes/pkg/api/legacyscheme"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	//add apps group schemes to legacyscheme
+	_ "k8s.io/kubernetes/pkg/apis/apps/install"
+	//add core group schemes to legacyscheme
+	_ "k8s.io/kubernetes/pkg/apis/core/install"
+)
+
+// This file holds all operations about varying resources in ResourceDistribution process;
+// If you want to add new-type resource, you only need to modify this file.
+
+const (
+	SourceResourceDistributionOfResource = "kruise.io/from-resourcedistribution"
+)
+
+// UnifiedResource abstracts all behaviors of Resource
+type UnifiedResource interface {
+	GetName() string
+	GetObject() *runtime.Object
+	GetObjectCopy() runtime.Object
+	GetObjectMeta() *metav1.ObjectMeta
+	GetGroupVersionKind() *schema.GroupVersionKind
+}
+
+// Resource holds specific object
+type Resource struct {
+	Object           runtime.Object
+	GroupVersionKind schema.GroupVersionKind
+}
+
+func (r *Resource) GetName() string {
+	switch resource := r.Object.(type) {
+	case *corev1.Secret:
+		return resource.Name
+	case *corev1.ConfigMap:
+		return resource.Name
+	default:
+		return ""
+	}
+}
+
+func (r *Resource) GetObjectMeta() *metav1.ObjectMeta {
+	switch resource := r.Object.(type) {
+	case *corev1.Secret:
+		return &resource.ObjectMeta
+	case *corev1.ConfigMap:
+		return &resource.ObjectMeta
+	default:
+		return nil
+	}
+}
+
+func (r *Resource) GetObject() *runtime.Object {
+	return &r.Object
+}
+
+func (r *Resource) GetObjectCopy() runtime.Object {
+	return r.Object.DeepCopyObject()
+}
+
+func (r *Resource) GetGroupVersionKind() *schema.GroupVersionKind {
+	return &r.GroupVersionKind
+}
+
+// MakeUnifiedResourceFromObject receives runtime.Object, returns UnifiedResource
+func MakeUnifiedResourceFromObject(resourceObject runtime.Object, fldPath *field.Path) (resource UnifiedResource, allErrs field.ErrorList) {
+	switch specificObject := resourceObject.(type) {
+	case *corev1.Secret, *corev1.ConfigMap:
+		resource = UnifiedResource(&Resource{
+			Object:           specificObject,
+			GroupVersionKind: specificObject.GetObjectKind().GroupVersionKind(),
+		})
+	default:
+		allErrs = append(allErrs, field.Invalid(fldPath, nil, "unsupported resource kind"))
+	}
+
+	return
+}
+
+// DeserializeResource receives yaml of resource, returns UnifiedResource
+func DeserializeResource(resourceYAML *runtime.RawExtension, fldPath *field.Path) (resource UnifiedResource, allErrs field.ErrorList) {
+	// check whether resource yaml is empty
+	if resourceYAML.Raw == nil {
+		allErrs = append(allErrs, field.Invalid(fldPath, nil, "empty resource is not allowed"))
+		return
+	}
+	// deserialize yaml
+	resourceObject, _, err := legacyscheme.Codecs.UniversalDeserializer().Decode(resourceYAML.Raw, nil, nil)
```

Comment:
Do not use `legacyscheme` because it is only for built-in types. You might need to use unstructured object.

Reply:
how about use a new scheme?
```
var (
	scheme = runtime.NewScheme()
	codecs = serializer.NewCodecFactory(scheme)
)

func init() {
	utilruntime.Must(corev1.AddToScheme(scheme))
}
```